lib/puppet/reports/http.rb

```diff
@@ -14,7 +14,7 @@
   def process
     url = URI.parse(Puppet[:reporturl])
     headers = { "Content-Type" => "application/x-yaml" }
-    options = {}
+    options = { :metric_id => [:puppet, :report, :http] }
```

Comment:
> I'm not sure what the implications would be if we suddenly started including :metric_id in all report posts on accident, but we probably don't want to?

We shouldn't ever end up with the case where elements passed in on the options hash are opaquely translated into content in the request, though, should we - e.g., as a query param on the URL or multipart form data?  My understanding has been that the options hash is a mechanism for passing through settings which must be explicitly read/acted upon by the connection class.

> It's not the cleanest thing but what do you think about checking if our Puppet::Network::HttpPool::connection_class is Puppet::Server::HttpClient instead of Puppet::Network::Http::Connection (or some other way of detecting that we're in the right environment) and only then adding :metric_id to the options?

That might be appropriate for the code change on this PR.  It's worth keeping in mind, though, that this client is used for various extension points whose source code lives outside of the core Ruby Puppet repo - e.g., in the PuppetDB and Node Classifier terminii.  We're adding the use of the `:metric-id` to those terminii.  Non-Puppet owned extensions may be using this as well.  Embedding knowledge about the specific implementation class used in code that lives outside of the Puppet repo would seem dangerous - e.g., in the event that a future refactor were to change the name of the default implementation class which gets used.  Also, while this code wouldn't add anything which would use the `:metric-id` when running outside of Puppet Server, I could envision a time in the future where having the `:metric-id` acted upon might be useful.  For example, in the `Puppet::Network::HTTP::Connection`, we might want to wrap a Puppet profiler block around the HTTP get/post calls and use the supplied `:metric-id` as an identifier for those.  This could be nice for profiling http client calls made by the agent - in a way that can work fairly generically for extensions which live outside of the Ruby Puppet source.

I agree that it does seem a little distant to have an option set which is only acted on via code which lives in another repo.  I wonder if a decent compromise may be to expand the [option](https://github.com/puppetlabs/puppet/blob/4.9.3/lib/puppet/network/http/connection.rb#L66-L67) comments in the connection.rb class to mention the `:metric-id` as a possible option but that it is not actually used by the `Puppet::Network::Http::Connection` implementation.  We could also mention in these comments that any options not recognized by the `Connection` class will be ignored, that no error will be thrown.  We could also add a comment here and where used in the reports.rb code that that the `:metric-id` option is being added for Puppet Server's use.

@MosesMendoza - WDYT?